src/shared/TestInfrastructure/Objects/TestGit.cs

```diff
@@ -8,36 +8,73 @@ namespace Microsoft.Git.CredentialManager.Tests.Objects
 {
     public class TestGit : IGit
     {
-        public TestGitConfiguration NoRepoConfiguration { get; } = new TestGitConfiguration();
+        public TestGit(IDictionary<string, string> globalConfig = null)
+        {
+            GlobalConfiguration = new TestGitConfiguration(globalConfig);
+        }
+
+        public TestGitConfiguration GlobalConfiguration { get; }
 
-        public IDictionary<string, TestGitConfiguration> Repositories { get; } =
-            new Dictionary<string, TestGitConfiguration>();
+        public IDictionary<string, TestGitRepository> Repositories { get; } =
+            new Dictionary<string, TestGitRepository>();
 
-        public TestGitConfiguration AddRepository(string repoPath, IDictionary<string, string> config = null)
+        public TestGitRepository AddRepository(string repoPath, IDictionary<string, string> config = null)
         {
-            var repoConfig = new TestGitConfiguration
-            {
-                RepositoryPath = repoPath,
-                Values = config ?? new Dictionary<string, string>()
-            };
+            var repoConfig = new TestGitConfiguration(repoPath, config);
+            var repo = new TestGitRepository(repoPath, repoConfig);
 
-            Repositories.Add(repoPath, repoConfig);
+            Repositories.Add(repoPath, repo);
 
-            return repoConfig;
+            return repo;
         }
 
         #region IGit
 
         void IDisposable.Dispose() { }
 
         IGitConfiguration IGit.GetConfiguration(string repositoryPath)
-            => string.IsNullOrWhiteSpace(repositoryPath)
-            ? NoRepoConfiguration
-            : Repositories[repositoryPath];
+        {
+            if (string.IsNullOrWhiteSpace(repositoryPath) || !Repositories.TryGetValue(repositoryPath, out TestGitRepository repo))
+            {
+                return GlobalConfiguration;
+            }
+
+            IDictionary<string, string> mergedConfigDict = MergeDictionaries(GlobalConfiguration.Dictionary, repo.Configuration.Dictionary);
+
+            return new TestGitConfiguration(repositoryPath, mergedConfigDict);
+        }
 
         string IGit.GetRepositoryPath(string path) =>
             Repositories.Keys.FirstOrDefault(path.StartsWith);
 
         #endregion
+
+        private static IDictionary<string, string> MergeDictionaries(params IDictionary<string, string>[] dictionaries)
```

Comment:
By duplicate variables do you mean ones that differ only by case?
Because the default dictionary constructor here is does case-sensitive hashing, we will allow multiple entries here.

@dscho are Git configuration entries case-sensitive or case-insensitive? For example are `git config http.proxy` and `git config http.PROXY` the same?

Reply:
Git configuration keys are case-insensitive, at least partially so. Therefore: yes, `http.proxy` and `http.PROXY` are considered equivalent.

Having said that, there is an exception to this rule: if you have more than one period in the key, the part between the first and the last one is _case-sensitive_.

For example, `http.https://GitHub.com.proxy` does _not_ refer to the same entry as `http.https://github.com.proxy`. Yes, I know. I know. Oh my, do I know...

Reply:
> Yes, I know. I know. Oh my, do I know...

I want to cry... 😢